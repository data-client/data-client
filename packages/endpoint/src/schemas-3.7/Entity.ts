/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

import { isImmutable, denormalizeImmutable } from './ImmutableUtils.js';
import type { UnvisitFunction } from './interface.js';
import { Entity as EntitySchema } from './schema.js';
import { AbstractInstanceType } from '../normal.js';

const EmptyBase = class {} as any as abstract new (...args: any[]) => {
  pk(parent?: any, key?: string): string | undefined;
};

/**
 * Represents data that should be deduped by specifying a primary key.
 * @see https://resthooks.io/docs/api/Entity
 */
export default abstract class Entity extends EntitySchema(EmptyBase) {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  abstract pk(parent?: any, key?: string): string | undefined;

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /** Return true to merge incoming data; false keeps existing entity
   *
   * @see https://resthooks.io/docs/api/schema.Entity#useIncoming
   */
  static useIncoming(
    existingMeta: { date: number; fetchedAt: number },
    incomingMeta: { date: number; fetchedAt: number },
    existing: any,
    incoming: any,
  ) {
    return existingMeta.fetchedAt <= incomingMeta.fetchedAt;
  }

  /** Run when an existing entity is found in the store */
  static mergeWithStore(
    existingMeta:
      | {
          date: number;
          fetchedAt: number;
        }
      | undefined,
    incomingMeta: { date: number; fetchedAt: number },
    existing: any,
    incoming: any,
  ) {
    const useIncoming =
      // we may have in store but not in meta; so this existance check is still important
      !existingMeta ||
      this.useIncoming(existingMeta, incomingMeta, existing, incoming);

    if (useIncoming) {
      // distinct types are not mergeable, so just replace
      if (typeof incoming !== typeof existing) {
        return incoming;
      } else {
        return this.merge(existing, incoming);
      }
    } else {
      return existing;
    }
  }

  /** Factory method to convert from Plain JS Objects.
   *
   * @param [props] Plain Object of properties to assign.
   */
  declare static fromJS: <T extends typeof Entity>(
    this: T,
    // TODO: this should only accept members that are not functions
    props?: Partial<AbstractInstanceType<T>>,
  ) => AbstractInstanceType<T>;

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   */
  declare static pk: <T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
  ) => string | undefined;

  /** Do any transformations when first receiving input */
  static process(input: any, parent: any, key: string | undefined): any {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      if (Array.isArray(input)) {
        const errorMessage = `Attempted to initialize ${
          this.name
        } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://resthooks.io/docs/guides/debugging
Learn more about schemas: https://resthooks.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
        if (this.automaticValidation !== 'warn') {
          const error = new Error(errorMessage);
          (error as any).status = 400;
          throw error;
        }
        console.warn(errorMessage);
      }
    }

    return super.process(input, parent, key);
  }

  static validate(processedEntity: any): string | undefined {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      const keysOfRecord = new Set(Object.keys(this.defaults));
      const keysOfProps = Object.keys(processedEntity);
      const [found, missing, unexpected] = [[], [], []] as [
        string[],
        string[],
        string[],
      ];
      for (const keyOfProps of keysOfProps) {
        if (keysOfRecord.has(keyOfProps)) {
          found.push(keyOfProps);
        } else {
          unexpected.push(keyOfProps);
        }
      }
      for (const keyOfRecord of keysOfRecord) {
        if (!found.includes(keyOfRecord)) {
          missing.push(keyOfRecord);
        }
      }

      // only bother with this if they used *any* defaults
      if (keysOfRecord.size) {
        const tooManyUnexpected =
          // unexpected compared to members in response
          Math.max(keysOfProps.length / 2, 1) <= unexpected.length &&
          // unexpected compared to what we specified
          keysOfRecord.size > Math.max(unexpected.length, 2) &&
          // as we find more and more be more easily assured it is correct
          found.length ** 1.5 / 2 <= unexpected.length;
        const foundNothing = found.length < Math.min(1, keysOfRecord.size / 2);
        // if we find nothing (we expect at least one member for a pk)
        // or we find too many unexpected members
        if (tooManyUnexpected || foundNothing) {
          let extra = '';
          let reason = 'substantially different than expected keys';
          if (foundNothing) {
            extra += `\n    Missing: ${missing}`;
            reason = 'no matching keys found';
          }
          if (tooManyUnexpected) {
            extra += `\n    Unexpected keys: ${unexpected}`;
            reason = 'a large number of unexpected keys found';
          }
          const errorMessage = `Attempted to initialize ${
            this.name
          } with ${reason}

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://resthooks.io/docs/guides/debugging
  Learn more about schemas: https://resthooks.io/docs/api/schema
  If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

  Expected keys:
    Found: ${found}${extra}
  Value (processed): ${JSON.stringify(processedEntity, null, 2)}`;
          if (
            (found.length >= 4 && tooManyUnexpected) ||
            this.automaticValidation === 'warn'
          ) {
            console.warn(errorMessage);
          } else {
            return errorMessage;
          }
        }
      }
    }
    return super.validate(processedEntity);
  }

  static denormalize<T extends typeof Entity>(
    this: T,
    input: any,
    unvisit: UnvisitFunction,
  ): [denormalized: AbstractInstanceType<T>, found: boolean, suspend: boolean] {
    // TODO: remove codecov ignore once denormalize is modified to expect this
    /* istanbul ignore if */
    if (typeof input === 'symbol') {
      return [undefined, true, true] as any;
    }
    // TODO(breaking): Remove fromJS and setLocal call once old versions are no longer supported
    if (isImmutable(input)) {
      if (this.validate((input as any).toJS()))
        return [undefined as any, false, true];
      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      unvisit.setLocal?.(input);
      const [denormEntity, found, deleted] = denormalizeImmutable(
        this.schema,
        input,
        unvisit,
      );
      return [this.fromJS(denormEntity.toObject()) as any, true, deleted];
    }
    let entityCopy: any;
    // new path
    if (input instanceof this) {
      entityCopy = input;
      // TODO(breaking): Remove fromJS and setLocal call once old versions are no longer supported
    } else {
      if (this.validate(input)) {
        return [undefined as any, false, true];
      }
      entityCopy = this.fromJS(input);
      // Need to set this first so that if it is referenced further within the
      // denormalization the reference will already exist.
      unvisit.setLocal?.(entityCopy);
    }

    let deleted = false;

    // note: iteration order must be stable
    Object.keys(this.schema).forEach(key => {
      const schema = this.schema[key];
      const nextInput = (input as any)[key];
      const [value, , deletedItem] = unvisit(nextInput, schema);

      if (deletedItem && !!this.defaults[key]) {
        deleted = true;
      }
      if ((input as any)[key] !== value) {
        // we're cheating because we know it is implemented
        (this as any).set(entityCopy, key, value);
      }
    });

    return [entityCopy, true, deleted];
  }

  /** Used by denormalize to set nested members */
  protected static set?(entity: any, key: string, value: any) {
    entity[key] = value;
  }
}

if (process.env.NODE_ENV !== 'production') {
  /* istanbul ignore else */
  const superFrom = Entity.fromJS;
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof Entity>(
    this: T,
    props?: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === Entity.prototype.pk)
      throw new Error('cannot construct on abstract types');
    return superFrom.call(this, props) as any;
  };
}
