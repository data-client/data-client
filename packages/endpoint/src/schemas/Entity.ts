/* eslint-disable @typescript-eslint/explicit-module-boundary-types */

import { AbstractInstanceType } from '../normal.js';
import { Entity as EntitySchema } from '../schema.js';

const EmptyBase = class {} as any as abstract new (...args: any[]) => {
  pk(parent?: any, key?: string, args?: readonly any[]): string | undefined;
};

/**
 * Represents data that should be deduped by specifying a primary key.
 * @see https://dataclient.io/docs/api/Entity
 */
export default abstract class Entity extends EntitySchema(EmptyBase) {
  /**
   * A unique identifier for each Entity
   *
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   * @param [args] ...args sent to Endpoint
   * @see https://dataclient.io/rest/api/Entity#pk
   */
  abstract pk(
    parent?: any,
    key?: string,
    args?: readonly any[],
  ): string | undefined;

  /** Control how automatic schema validation is handled
   *
   * `undefined`: Defaults - throw error in worst offense
   * 'warn': only ever warn
   * 'silent': Don't bother with processing at all
   *
   * Note: this only applies to non-nested members.
   */
  protected declare static automaticValidation?: 'warn' | 'silent';

  /** Factory method to convert from Plain JS Objects.
   *
   * @param [props] Plain Object of properties to assign.
   * @see https://dataclient.io/rest/api/Entity#fromJS
   */
  declare static fromJS: <T extends typeof Entity>(
    this: T,
    // TODO: this should only accept members that are not functions
    props?: Partial<AbstractInstanceType<T>>,
  ) => AbstractInstanceType<T>;

  /**
   * A unique identifier for each Entity
   *
   * @param [value] POJO of the entity or subset used
   * @param [parent] When normalizing, the object which included the entity
   * @param [key] When normalizing, the key where this entity was found
   * @param [args] ...args sent to Endpoint
   */
  declare static pk: <T extends typeof Entity>(
    this: T,
    value: Partial<AbstractInstanceType<T>>,
    parent?: any,
    key?: string,
    args?: any[],
  ) => string | undefined;

  /** Do any transformations when first receiving input
   *
   * @see https://dataclient.io/docs/api/Entity#process
   */
  static process(
    input: any,
    parent: any,
    key: string | undefined,
    args: any[],
  ): any {
    /* istanbul ignore else */
    if (
      process.env.NODE_ENV !== 'production' &&
      this.automaticValidation !== 'silent'
    ) {
      if (Array.isArray(input)) {
        const errorMessage = `Attempted to initialize ${
          this.name
        } with an array, but named members were expected

This is likely due to a malformed response.
Try inspecting the network response or fetch() return value.
Or use debugging tools: https://dataclient.io/docs/guides/debugging
Learn more about schemas: https://dataclient.io/docs/api/schema
If this is a mistake, you can disable this check by setting static automaticValidation = 'silent'

First three members: ${JSON.stringify(input.slice(0, 3), null, 2)}`;
        if (this.automaticValidation !== 'warn') {
          const error = new Error(errorMessage);
          (error as any).status = 400;
          throw error;
        }
        console.warn(errorMessage);
      }
    }

    return super.process(input, parent, key, args);
  }

  declare static denormalize: <T extends typeof Entity>(
    this: T,
    input: any,
    args: readonly any[],
    unvisit: (input: any, schema: any) => any,
  ) => AbstractInstanceType<T>;
}

if (process.env.NODE_ENV !== 'production') {
  /* istanbul ignore else */
  const superFrom = Entity.fromJS;
  // for those not using TypeScript this is a good catch to ensure they are defining
  // the abstract members
  Entity.fromJS = function fromJS<T extends typeof Entity>(
    this: T,
    props?: Partial<AbstractInstanceType<T>>,
  ): AbstractInstanceType<T> {
    if ((this as any).prototype.pk === Entity.prototype.pk)
      throw new Error('cannot construct on abstract types');
    return superFrom.call(this, props) as any;
  };
}
