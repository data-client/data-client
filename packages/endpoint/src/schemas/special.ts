export const CREATE = Symbol('create');
