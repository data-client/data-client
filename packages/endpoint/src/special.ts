export const DELETED = Symbol('ENTITY WAS DELETED');
export const INVALID = DELETED;
