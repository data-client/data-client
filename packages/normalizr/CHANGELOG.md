# Change Log

## 0.10.0

### Minor Changes

- [#2912](https://github.com/reactive/data-client/pull/2912) [`922be79`](https://github.com/reactive/data-client/commit/922be79169a3eeea8e336eee519c165431ead474) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGE: `null` inputs are no longer filtered from Array or Object

  - `[]` and [schema.Array](https://dataclient.io/rest/api/Array) now behave in the same manner.
  - `null` values are now consistently handled everywhere (being retained).
    - These were already being retained in [nested Entities](https://dataclient.io/rest/guides/relational-data#nesting)
  - `undefined` is still filtered out.

### Patch Changes

- [`67f4e0b`](https://github.com/reactive/data-client/commit/67f4e0b45068da32d20e250267cb1cd2cea51226) Thanks [@ntucker](https://github.com/ntucker)! - Update README

- [`053e823`](https://github.com/reactive/data-client/commit/053e82377bd29f200cd7dfbc700da7a3ad7fa8d7) Thanks [@ntucker](https://github.com/ntucker)! - Update NextJS Demo link

## 0.9.5

### Patch Changes

- [`bb4b9583c5`](https://github.com/reactive/data-client/commit/bb4b9583c52e2b2fe45765af10b385b571901ee7) Thanks [@ntucker](https://github.com/ntucker)! - docs: Update readme

## 0.9.4

### Patch Changes

- [`d1b51af7ac`](https://github.com/reactive/data-client/commit/d1b51af7ac4a8a7c0559f478cc9503be8e61514c) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles by ensuring dependencies are built in order

## 0.9.3

### Patch Changes

- [#2818](https://github.com/reactive/data-client/pull/2818) [`fc0092883f`](https://github.com/reactive/data-client/commit/fc0092883f5af42a5d270250482b7f0ba9845e95) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles and update names

  - Client packages namespace into RDC
    - @data-client/react - RDC
    - @data-client/core - RDC.Core
    - @data-client/redux - RDC.Redux
  - Definition packages namespace top level
    - @data-client/rest - Rest
    - @data-client/graphql - GraphQL
    - @data-client/img - Img
    - @data-client/endpoint - Endpoint
  - Utility
    - @data-client/normalizr - normalizr
    - @data-client/use-enhanced-reducer - EnhancedReducer

## 0.9.2

### Patch Changes

- [`4ea0bc83f6`](https://github.com/reactive/data-client/commit/4ea0bc83f65f49cb2155f6aecdc5f8d1b168fd5e) Thanks [@ntucker](https://github.com/ntucker)! - Docs: Update repo links to reactive organization

## 0.8.0

### Minor Changes

- [#2784](https://github.com/reactive/data-client/pull/2784) [`c535f6c0ac`](https://github.com/reactive/data-client/commit/c535f6c0ac915b5242c1c7694308b7ee7aab16a1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGES:

  - DELETE removed -> INVALIDATE
  - drop all support for legacy schemas
    - entity.expiresAt removed
    - Collections.infer does entity check
    - all Entity overrides for backcompat are removed - operates just like EntitySchema, except with extra validation

- [#2795](https://github.com/reactive/data-client/pull/2795) [`79e286109b`](https://github.com/reactive/data-client/commit/79e286109b5566f8e7acfdf0f44201263072d1d1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: [Schema Serializers](https://dataclient.io/rest/guides/network-transform#deserializing-fields) _must_ support function calls

  This means Date will no longer work like before. Possible migrations:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: Temporal.Instant.from,
    };
  }
  ```

  or to continue using Date:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: (iso: string) => new Date(iso),
    };
  }
  ```

- [#2792](https://github.com/reactive/data-client/pull/2792) [`35ccedceb5`](https://github.com/reactive/data-client/commit/35ccedceb53d91dd54dd996990c7c75719be2b85) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: Serializer schemas are only processed during denormalization

### Patch Changes

- [#2779](https://github.com/reactive/data-client/pull/2779) [`ff51e71f45`](https://github.com/reactive/data-client/commit/ff51e71f45857eb172f3fe05829e34c9abb68252) Thanks [@ntucker](https://github.com/ntucker)! - Update jsdocs references to dataclient.io

## 0.2.2

### Patch Changes

- 7b835f113a: Improve package tags

## 0.2.1

### Patch Changes

- 954e06e839: docs: Update README

## 0.2.0

### Minor Changes

- bf141cb5a5: Removed deprecated Endpoint.optimisticUpdate -> use Endpoint.getOptimisticResponse

### Patch Changes

- 87475a0cae: Mark expiresAt and useIncoming as deprecated in the public interface (still fine internally)
