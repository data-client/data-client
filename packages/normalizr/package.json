{
  "name": "@data-client/normalizr",
  "version": "0.9.5",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://dataclient.io/docs/concepts/normalization",
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/normalizr"
  },
  "keywords": [
    "normalize",
    "denormalize",
    "flux",
    "redux",
    "data",
    "model",
    "cache",
    "api",
    "react-native",
    "ios",
    "android",
    "web",
    "schema",
    "typescript",
    "validation",
    "relational",
    "reactive",
    "json"
  ],
  "files": [
    "dist/",
    "lib/",
    "node.mjs",
    "legacy",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "sideEffects": false,
  "scripts": {
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development yarn g:rollup",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production yarn g:rollup",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean",
    "build:legacy-types": "yarn g:downtypes lib ts3.4",
    "prepublishOnly": "run build"
  },
  "author": "Nathaniel Tucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@types/node": "^20.0.0",
    "immutable": "4.2.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.17.0"
  }
}
