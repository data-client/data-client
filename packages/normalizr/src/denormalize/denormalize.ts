import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { Denormalize, DenormalizeNullable } from '../types.js';
import { getEntities } from '../WeakEntityMap.js';

export const denormalize = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
): Denormalize<S> | DenormalizeNullable<S> => {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(getEntities(entities), new LocalCache())(input, schema)[0];
};
