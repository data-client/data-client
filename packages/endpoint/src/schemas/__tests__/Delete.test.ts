// eslint-env jest
import { Schema, normalize } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';
import { fromJS } from 'immutable';

import { denormalizeSimple, denormalizeLegacy } from './denormalize';
import { schema } from '../../';
import { DELETED } from '../../special';
import Entity from '../Entity';

let dateSpy: jest.SpyInstance;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe(`${schema.Delete.name} normalization`, () => {
  test('throws if not given an entity', () => {
    // @ts-expect-error
    expect(() => new schema.Delete()).toThrow();
  });

  test('normalizes an object', () => {
    class User extends IDEntity {}

    expect(
      normalize({ id: '1', type: 'users' }, new schema.Delete(User)),
    ).toMatchSnapshot();
  });

  test('normalizes already processed entities', () => {
    class MyEntity extends IDEntity {}
    expect(normalize('1', new schema.Delete(MyEntity))).toMatchSnapshot();
    expect(
      normalize(['1', '2'], new schema.Array(new schema.Delete(MyEntity))),
    ).toMatchSnapshot();
  });

  test('does not infer', () => {
    class User extends IDEntity {}

    expect(
      new schema.Delete(User).infer([{ id: 5 }], {}, () => undefined),
    ).toBeUndefined();
  });

  it('should throw a custom error if data does not include pk', () => {
    class MyEntity extends Entity {
      readonly name: string = '';
      readonly secondthing: string = '';
      pk() {
        return this.name;
      }
    }
    function normalizeBad() {
      normalize({ secondthing: 'hi' }, new schema.Delete(MyEntity));
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });
});

describe(`${schema.Delete.name} denormalization`, () => {
  class User extends IDEntity {
    readonly username: string = '';
    readonly type: string = '';
  }
  const entities = {
    User: {
      '1': User.fromJS({ id: '1', username: 'Janey', type: 'users' }),
    },
  };

  test('denormalizes an object in the same manner as the Entity', () => {
    const user = denormalizeSimple('1', new schema.Delete(User), entities);
    expect(user).not.toEqual(expect.any(Symbol));
    if (typeof user === 'symbol') return;
    expect(user).toBeDefined();
    expect(user).toBeInstanceOf(User);
    expect(user?.username).toBe('Janey');
  });

  describe.each([
    ['direct', <T>(data: T) => data],
    ['immutable', fromJS],
  ])(`input (%s)`, (_, createInput) => {
    describe.each([
      [
        'class',
        <T extends Schema>(sch: T) => new schema.Array(sch),
        <T extends Record<string, any>>(sch: T) => new schema.Object(sch),
        denormalizeSimple,
      ],
      [
        'class, legacy',
        <T extends Schema>(sch: T) => new schema.Array(sch),
        <T extends Record<string, any>>(sch: T) => new schema.Object(sch),
        denormalizeLegacy,
      ],
      [
        'object, direct',
        <T extends Schema>(sch: T) => [sch],
        <T extends Record<string, any>>(sch: T) => sch,
        denormalizeSimple,
      ],
    ])(
      `schema construction (%s)`,
      (_, createArray, createObject, denormalize) => {
        test('denormalizes deleted entities as symbol', () => {
          const user = denormalize(
            '1',
            new schema.Delete(User),
            createInput({
              User: { '1': DELETED },
            }),
          );
          expect(user).toEqual(expect.any(Symbol));

          expect(
            denormalize(
              createInput({ data: '1' }),
              createObject({ data: new schema.Delete(User) }),
              createInput({
                User: { '1': DELETED },
              }),
            ),
          ).toEqual(expect.any(Symbol));
        });

        test('denormalize removes deleted entries in array', () => {
          expect(
            denormalize(
              createInput([{ data: '1' }]),
              createArray(createObject({ data: new schema.Delete(User) })),
              createInput({
                User: { '1': DELETED },
              }),
            ),
          ).toMatchSnapshot();
          expect(
            denormalize(
              createInput([{ data: '1' }]),
              createArray(createObject({ data: User })),
              createInput({
                User: { '1': DELETED },
              }),
            ),
          ).toMatchSnapshot();
        });

        test('denormalize sets undefined entities that are not present', () => {
          expect(
            denormalize(
              createInput([{ data: '1' }]),
              createArray(createObject({ data: new schema.Delete(User) })),
              createInput([{}]),
            ),
          ).toMatchSnapshot();

          expect(
            denormalize(
              createInput([{ data: '1' }]),
              createArray(createObject({ data: User })),
              createInput([{}]),
            ),
          ).toMatchSnapshot();

          expect(
            denormalize(
              createInput({ data: '1' }),
              createObject({ data: User }),
              createInput({}),
            ),
          ).toMatchSnapshot();
        });
      },
    );
  });
});
