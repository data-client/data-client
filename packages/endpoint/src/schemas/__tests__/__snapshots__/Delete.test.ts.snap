// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete denormalization input (direct) schema construction (class) denormalize removes deleted entries in array 1`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (class) denormalize removes deleted entries in array 2`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (class) denormalize sets undefined entities that are not present 1`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (class) denormalize sets undefined entities that are not present 2`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (class) denormalize sets undefined entities that are not present 3`] = `
{
  "data": undefined,
}
`;

exports[`Delete denormalization input (direct) schema construction (class, legacy) denormalize removes deleted entries in array 1`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (class, legacy) denormalize removes deleted entries in array 2`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (class, legacy) denormalize sets undefined entities that are not present 1`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (class, legacy) denormalize sets undefined entities that are not present 2`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (class, legacy) denormalize sets undefined entities that are not present 3`] = `
{
  "data": undefined,
}
`;

exports[`Delete denormalization input (direct) schema construction (object, direct) denormalize removes deleted entries in array 1`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (object, direct) denormalize removes deleted entries in array 2`] = `[]`;

exports[`Delete denormalization input (direct) schema construction (object, direct) denormalize sets undefined entities that are not present 1`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (object, direct) denormalize sets undefined entities that are not present 2`] = `
[
  {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (direct) schema construction (object, direct) denormalize sets undefined entities that are not present 3`] = `
{
  "data": undefined,
}
`;

exports[`Delete denormalization input (immutable) schema construction (class) denormalize removes deleted entries in array 1`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (class) denormalize removes deleted entries in array 2`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (class) denormalize sets undefined entities that are not present 1`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (class) denormalize sets undefined entities that are not present 2`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (class) denormalize sets undefined entities that are not present 3`] = `
Immutable.Map {
  "data": undefined,
}
`;

exports[`Delete denormalization input (immutable) schema construction (class, legacy) denormalize removes deleted entries in array 1`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (class, legacy) denormalize removes deleted entries in array 2`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (class, legacy) denormalize sets undefined entities that are not present 1`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (class, legacy) denormalize sets undefined entities that are not present 2`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (class, legacy) denormalize sets undefined entities that are not present 3`] = `
Immutable.Map {
  "data": undefined,
}
`;

exports[`Delete denormalization input (immutable) schema construction (object, direct) denormalize removes deleted entries in array 1`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (object, direct) denormalize removes deleted entries in array 2`] = `Immutable.List []`;

exports[`Delete denormalization input (immutable) schema construction (object, direct) denormalize sets undefined entities that are not present 1`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (object, direct) denormalize sets undefined entities that are not present 2`] = `
Immutable.List [
  Immutable.Map {
    "data": undefined,
  },
]
`;

exports[`Delete denormalization input (immutable) schema construction (object, direct) denormalize sets undefined entities that are not present 3`] = `
Immutable.Map {
  "data": undefined,
}
`;

exports[`Delete normalization normalizes already processed entities 1`] = `
{
  "entities": {},
  "entityMeta": {},
  "indexes": {},
  "result": "1",
}
`;

exports[`Delete normalization normalizes already processed entities 2`] = `
{
  "entities": {},
  "entityMeta": {},
  "indexes": {},
  "result": [
    "1",
    "2",
  ],
}
`;

exports[`Delete normalization normalizes an object 1`] = `
{
  "entities": {
    "User": {
      "1": Symbol(ENTITY WAS DELETED),
    },
  },
  "entityMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`Delete normalization should throw a custom error if data does not include pk (serializes pk) 1`] = `
"Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://dataclient.io/docs/guides/debugging
  Learn more about schemas: https://dataclient.io/docs/api/schema

  Invalidate(Entity): Invalidate(MyEntity)
  Value (processed): {
  "secondthing": "hi"
}
  "
`;

exports[`Delete normalization should throw a custom error if data does not include pk 1`] = `
"Missing usable primary key when normalizing response.

  This is likely due to a malformed response.
  Try inspecting the network response or fetch() return value.
  Or use debugging tools: https://dataclient.io/docs/guides/debugging
  Learn more about schemas: https://dataclient.io/docs/api/schema

  Invalidate(Entity): Invalidate(MyEntity)
  Value (processed): {
  "secondthing": "hi"
}
  "
`;
